src/mail_data.c

```diff
@@ -0,0 +1,182 @@
+#include "global.h"
+#include "text.h"
+#include "mail.h"
+#include "mail_data.h"
+#include "constants/species.h"
+#include "constants/items.h"
+#include "pokemon_icon.h"
+
+void ClearMailData(void)
+{
+    u8 i;
+
+    for (i = 0; i < MAIL_COUNT; i++)
+        ClearMailStruct(&gSaveBlock1Ptr->mail[i]);
+}
+
+void ClearMailStruct(struct MailStruct *mail)
+{
+    s32 i;
+
+    for (i = 0; i < MAIL_WORDS_COUNT; i++)
+        mail->words[i] = 0xFFFF;
+    for (i = 0; i < PLAYER_NAME_LENGTH; i++)
+        mail->playerName[i] = EOS;
+    for (i = 0; i < 4; i++)
+        mail->trainerId[i] = 0;
+    mail->species = SPECIES_BULBASAUR;
+    mail->itemId = ITEM_NONE;
+}
+
+bool8 MonHasMail(struct Pokemon *mon)
+{
+    u16 heldItem = GetMonData(mon, MON_DATA_HELD_ITEM);
+    if (ItemIsMail(heldItem) && GetMonData(mon, MON_DATA_MAIL) != 0xFF)
+        return TRUE;
+    else
+        return FALSE;
+}
+
+u8 GiveMailToMon(struct Pokemon *mon, u16 itemId)
+{
+    u8 heldItem[2];
+    u8 id, i;
+    u16 species;
+    u32 personality;
+
+    heldItem[0] = itemId;
+    heldItem[1] = itemId >> 8;
+    for (id = 0; id < PARTY_SIZE; id++)
+    {
+        if (gSaveBlock1Ptr->mail[id].itemId == 0)
+        {
+            for (i = 0; i < MAIL_WORDS_COUNT; i++)
+                gSaveBlock1Ptr->mail[id].words[i] = 0xFFFF;
+            for (i = 0; i < PLAYER_NAME_LENGTH - 1 && gSaveBlock2Ptr->playerName[i] != EOS; i++)
+                gSaveBlock1Ptr->mail[id].playerName[i] = gSaveBlock2Ptr->playerName[i];
+            for (; i <= 5; i++)
+            	gSaveBlock1Ptr->mail[id].playerName[i] = CHAR_SPACE;
+            gSaveBlock1Ptr->mail[id].playerName[i] = EOS;
+            for (i = 0; i < 4; i++)
+                gSaveBlock1Ptr->mail[id].trainerId[i] = gSaveBlock2Ptr->playerTrainerId[i];
+            species = GetBoxMonData(&mon->box, MON_DATA_SPECIES);
+            personality = GetBoxMonData(&mon->box, MON_DATA_PERSONALITY);
+            gSaveBlock1Ptr->mail[id].species = SpeciesToMailSpecies(species, personality);
+            gSaveBlock1Ptr->mail[id].itemId = itemId;
+            SetMonData(mon, MON_DATA_MAIL, &id);
+            SetMonData(mon, MON_DATA_HELD_ITEM, heldItem);
+            return id;
+        }
+    }
+    return 0xFF;
+}
+
+u16 SpeciesToMailSpecies(u16 species, u32 personality)
```

Comment:
```c
u16 SpeciesToMailSpecies(u16 species, u32 personality)
{
    if (species == SPECIES_UNOWN)
    {
        species = GetUnownLetterByPersonality(personality);
        species += 1875;
    }
    return species;
}
```

Reply:
It doesn't work. I can use species macro tho

Reply:
Try 30000 instead of 1875, I miscounted the 0s.

Reply:
Still no luck. They actually optimized the function in EM is the word 0x7530 is used instead of 0x7530 << 16

Reply:
```c

u16 SpeciesToMailSpecies(u16 species, u32 personality)
{
    if (species == SPECIES_UNOWN) {
        u32 mailSpecies = GetUnownLetterByPersonality(personality) + 30000;
        return mailSpecies;
    }
    return species;
}
```